sycl/include/CL/sycl/intel/bit_cast.hpp

```diff
@@ -0,0 +1,44 @@
+//==----------- bit_cast.hpp --- SYCL bit_cast -----------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <CL/sycl/detail/defines.hpp>
+#include <CL/sycl/detail/helpers.hpp>
+
+#include <numeric>
+#include <type_traits>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+
+template <typename To, typename From>
+constexpr To bit_cast(const From &from) noexcept {
+  static_assert(sizeof(To) == sizeof(From),
+                "Sizes of To and From must be equal");
+  static_assert(std::is_trivially_copyable<From>::value,
+                "From must be trivially copyable");
+  static_assert(std::is_trivially_copyable<To>::value,
+                "To must be trivially copyable");
+#if __cpp_lib_bit_cast
+  return std::bit_cast<To>(from);
+#else // __cpp_lib_bit_cast
+
+#if __has_builtin(__builtin_bit_cast)
+  return __builtin_bit_cast(To, from);
+#else  // __has_builtin(__builtin_bit_cast)
+  To to;
```

Comment:
`To` must be trivially-constructible in this case.

Reply:
Good comment. However, if I understand correctly if we are using `clang` as a compiler, we should never fallback to this version.

Reply:
> we should never fallback to this version

We should emit an error here, then. Still, we implement some behaviour for this case. Hence, I reckon, `static_assert(std::is-trivially_constructible<To>::value, "Some good reasoning message");` is much more user-friendly than a generic compile-time error saying "your data type ain't got no viable constructor for me".

Reply:
Thanks, fixed.